pandas/core/arrays/categorical.py

```diff
@@ -323,14 +323,6 @@ def __init__(self, values, categories=None, ordered=None, dtype=None,
         # we may have dtype.categories be None, and we need to
         # infer categories in a factorization step futher below
 
-        if is_categorical(values):
-            # GH23814, for perf, if values._values already an instance of
-            # Categorical, set values to codes, and run fastpath
-            if (isinstance(values, (ABCSeries, ABCIndexClass)) and
-               isinstance(values._values, type(self))):
```

Comment:
Maybe we can keep this fastpath, but check that the `dtype` is equal or that `categories` and `ordered` are not set? 

So that the simple case of `pd.Categorical(Series[category])` still follows the fastpath.

Reply:
Ah, but I see you added a check for equal categories below, so that might ensure this is still fast enough (could you do a timing of the basic case with master vs your PR?)

Reply:
There are some benchmarks in the discussion #23814 which suggest that this approach is fast enough. Before that PR, the benchmark took 25.2ms and after the change in this PR, it sped up to 5.38ms. With the (broken) `ABCSeries` test, it takes only 3.13ms.